### PR TITLE
Add a CILViewer2D to the web app

### PR DESCRIPTION
**Pull request by Pasarus**
To be reviewed after: #249 and #255
Do not merge first

You need a python environment with Trame, and Pywebvue installed, alongside the normal requirements for CILViewer. You need to install it will pip, and ensure it is not the v2 release candidates.

I have been running this in a conda environment python version 3.9, I recommend you do the same.

To run the `CILViewer/Wrappers/Python/ccpi/web_viewer/web_app.py` file. You need to pass files or a directory that contains files to load. To run the 2D viewer instead of the 3D viewer pass either `-d` or `--2D`.

For example in the C`ILViewer/Wrappers/Python/ccpi/web_viewer directory`, run:
`python web_app.py --2D CILViewer/Wrappers/Python/ccpi/web_viewer/data`
This should launch the Trame/Web viewer application.

### CODE REVIEW

File: Wrappers/Python/ccpi/web_viewer/README.md
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/web_viewer/README.md:9]: you could add info on your entry point here

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/README.md:9]: I found --2D had to go before the path to the data, please can you make that clear here

**Review by lauramurgatroyd — CHANGES_REQUESTED** [review]
Hi, I am going to review twice - this review is based on me running the 2D viewer and what i found, the second review will be after I've looked at the code.

It is looking good but I found some bugs with the window/level...
- When I load the app, the window/level looks good and the slider looks correct
image

- When I press auto window-level this happens:
image

- After messing around with various (all of the) settings, if I click " reset defaults", it looks like this:
image
So again, different to the above 2.

I noticed the tracing is really laggy - is it similar for you?

I don't see an option to show the slice histogram - is that expected?

Do we have plans to add the line profile?

I tried to load a nexus file and got this error:

> ERROR:root:Exception raised
> ERROR:root:KeyError('opacity')
> ERROR:root:Traceback (most recent call last):
>   File "D:\lhe97136\Miniconda3\envs\cilviewer_webapp\lib\site-packages\wslink\backends\aiohttp\__init__.py", line 411, in onMessage
>     results = func(*args, **kwargs)
>   File "D:\lhe97136\Miniconda3\envs\cilviewer_webapp\lib\site-packages\pywebvue\protocol.py", line 275, in update_state
>     self.app.set(
>   File "D:\lhe97136\Miniconda3\envs\cilviewer_webapp\lib\site-packages\pywebvue\utils\__init__.py", line 230, in __exit__
>     fn(**self._app.state)
>   File "D:\lhe97136\Work\CCPi\CILViewer\Wrappers\Python\ccpi\web_viewer\web_app.py", line 113, in change_model
>     TRAME_VIEWER.load_file(kwargs['file_name'], windowing_method=kwargs['opacity'])
> KeyError: 'opacity'
>

File: Wrappers/Python/ccpi/viewer/CILViewer2D.py
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/viewer/CILViewer2D.py:2119]: This is not giving the map window for the slice, it is giving it for the whole 3D dataset.

Comment (Pasarus) [reply Wrappers/Python/ccpi/viewer/CILViewer2D.py:2119]: So this is just a bad name for the function, effectively it's doing the same thing as the Volume namesake in the none-2D viewer, it's been renamed to slice here because it's 2D alone. Perhaps I should rename it to something like `getAllSliceMapWindow`?

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/viewer/CILViewer2D.py:2119]: Currently in the 2D viewer we adjust the window/level based on the values in the current slice, not in the whole image. So auto window-level just takes into account the values on that slice, rather than in the whole volume. In the 3D viewer it's based on the whole volume.

We want the 2D and 3D viewer to perform in the same way for the slice, so the question is which do we want them both to do?
@paskino what do you think?
This can be resolved in a separate issue/PR though

File: Wrappers/Python/ccpi/viewer/CILViewer2D.py
Comment (Pasarus) [reply Wrappers/Python/ccpi/viewer/CILViewer2D.py:2121]: Removed!

File: Wrappers/Python/ccpi/web_viewer/trame_viewer2D.py
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/web_viewer/trame_viewer2D.py]: Typo in "reposition"

Comment (Pasarus) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer2D.py]: Fixed!

File: Wrappers/Python/ccpi/viewer/CILViewer2D.py
Comment (Pasarus) [reply Wrappers/Python/ccpi/viewer/CILViewer2D.py:2122]: It gives the raw value (867, 867)

File: Wrappers/Python/ccpi/web_viewer/trame_viewer.py
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/web_viewer/trame_viewer.py:378]: i don't think this should be here

Comment (Pasarus) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:378]: Ya know I think the merge broke a couple things, let me clean this up!
```suggestion
```

File: Wrappers/Python/ccpi/viewer/CILViewer.py
Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/viewer/CILViewer.py:1157]: I suggest we rename setSliceColorWindowLevelPercentiles -> setSliceColorWindowPercentiles

We also need to fix the behaviour of the method - I will make a separate PR for it

Comment (Pasarus) [reply Wrappers/Python/ccpi/viewer/CILViewer.py:1157]: Just to confirm this is not required to change in this PR, as it was not made or renamed in this PR, but a later one?

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/viewer/CILViewer.py:1157]: yes, will be fixed separately, issue is here https://github.com/vais-ral/CILViewer/issues/258

File: Wrappers/Python/ccpi/web_viewer/trame_viewer.py
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: why is this 60 and 95?

File: Wrappers/Python/ccpi/web_viewer/trame_viewer.py
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/web_viewer/trame_viewer.py:267]: The slice window range defaults should not come from the VolumeMapWindow

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:267]: We should replace this whole if/else with:

```python
ia = vtk.vtkImageHistogramStatistics()
ia.SetInputData(self.cil_viewer.voi.GetOutput())
ia.SetAutoRangePercentiles(0, 100)
ia.Update()
self.cmin, self.cmax = ia.GetAutoRange()
ia.SetAutoRangePercentiles(5.0, 95.)
ia.Update()
self.slice_window_range_defaults = ia.GetAutoRange()
ia.Update()
```

What this does is it sets cmin and cmax to the min and max values for the SLICE. In the 2D viewer that is all you need. In the 3D viewer, these are different to the min and max value for the entire image volume.

To get the min and max value for the image volume you need:
`cmin, cmax = self.cil_viewer.getVolumeMapWindow((0., 100.), "scalar") `, like you have at line 266.

So what you will need for the 3D viewer is to store two different sets of cmin and cmax, one which is for the slice slider and one which is for the volume slider. And the slice slider range needs updating whenever the slice number is changed.

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:267]: What I will do later is add the code for getting the cmin and cmax for the slice and getting the window range defaults for the slice as methods in both the 2D and 3D viewer (with the same names), so that later we can call these instead of you having to use all this code in the web app.

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:267]: But these additions to the viewer will be done after this PR is merged, just to clarify.

**Review by lauramurgatroyd — CHANGES_REQUESTED** [review]

File: Wrappers/Python/ccpi/web_viewer/trame_viewer.py
Comment (Pasarus) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: This is close to the default values that are being used by the CILViewer2D I can find no reason why they are, I got to these numbers through trial and error and comparison with the (60. 95.) and default from the 2D viewer.

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: That's weird, because the 2D viewer sets the percentiles here:
https://github.com/vais-ral/CILViewer/blob/47f45a970d6931d34173c5418737f765ab83658b/Wrappers/Python/ccpi/viewer/CILViewer2D.py#L1625

Comment (Pasarus) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: Yeah I know, it's a very weird behaviour that I can't find the source for.

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: what happens when you set 5, 95?

Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: problem could've been related to incorrect window definition in viewer?

Comment (Pasarus) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: The min value was 0. up until about 60% but the viewer was setting it to about 130-150. which is roughly what we get when we use 60% as the min value.

File: Wrappers/Python/ccpi/web_viewer/trame_viewer.py
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/web_viewer/trame_viewer.py:378]:
```suggestion
            self.cil_viewer.updatePipeline()
```

File: Wrappers/Python/ccpi/web_viewer/trame_viewer.py
Comment (lauramurgatroyd) [reply Wrappers/Python/ccpi/web_viewer/trame_viewer.py:264]: Part of the problem is the getSliceMapWindow method isn't getting the window for the slice (as previously discussed). So 5, 95 will be correct when the other issues are fixed.

File: Wrappers/Python/ccpi/web_viewer/trame_viewer2D.py
Comment (lauramurgatroyd) [inline Wrappers/Python/ccpi/web_viewer/trame_viewer2D.py:158]:
```suggestion
        window = cmax - cmin
```

**Comment by lauramurgatroyd** [comment]
We still have some small issues with window/level and some things we need to discuss re: this. I will make issues to cover these.

Otherwise, all working and great work 👍 :)

**Pull request merged**
